### ISSUE TITLE
Error IMAP4Client.fetchSpecific an HTML body part

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4049]
|<img alt="biny's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| biny reported|
|-|-|
|Trac ID|trac#4049|
|Type|defect|
|Created|2009-10-05 13:47:49Z|
|Branch|https://github.com/twisted/twisted/tree/imap-fetch-specific-html-4049|

New test for twisted.mail.test.test_imap.IMAP4ClientFetchTests:

```
    def test_fetchSpecificHTML(self):
        d = self.client.fetchSpecific('7')
        self.assertEquals(
            self.transport.value(), '0001 FETCH 7 BODY[]\r\n')
        self.client.lineReceived('* 7 FETCH (BODY[] "<html>test</html>")')
        self.client.lineReceived('0001 OK FETCH completed')
        self.assertEquals(
            self._extractDeferredResult(d), {7: [['BODY', [], "<html>test</html>"]]})
```

```
[ERROR]: twisted.mail.test.test_imap.IMAP4ClientFetchTests.test_fetchSpecificHTML

Traceback (most recent call last):
  File "/home/biny/packages/repositories/twisted/twisted/mail/imap4.py", line 3419, in _cbFetch
    mapping = self._parseFetchPairs(values[0])
  File "/home/biny/packages/repositories/twisted/twisted/mail/imap4.py", line 3401, in _parseFetchPairs
    "Not enough arguments", fetchResponseList)
twisted.mail.imap4.IllegalServerResponse: ('Not enough arguments', ['BODY', [], '<html>test</html>'])
```

Attachments:

* [imap4-parser-changes.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b1d/b1d49039967f7fd493ea6ebac0b5dabbc86e37bb/57cae8b35b4a0ebe19befa0489cb0ec28d972355.patch) (11743 bytes) - added by exarkun on 2009-11-19 21:05:32Z - Some stuff I did, not being proposed to be committed anyplace, just sayin'

<details><summary>Searchable metadata</summary>

```
trac-id__4049 4049
type__defect defect
reporter__biny biny
priority__normal normal
milestone__ 
branch__branches_imap_fetch_specific_html_4049 branches/imap-fetch-specific-html-4049
branch_author__biny__exarkun biny, exarkun
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1254750469000000 1254750469000000
changetime__1259902834000000 1259902834000000
version__None None
owner__ 
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_2"></a>|@thijstriemstra set owner to biny|
|-|-|

Thanks, can you provide a unified diff against the twisted trunk? this makes it possible for us to test this change against many platforms using our [build farm](http://buildbot.twistedmatrix.com).

**Comment by twisted-trac** [comment]
|<img alt="biny's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|biny set owner to @thijstriemstra|
|-|-|

There isn't any change here. Only a new failing test.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_4"></a>|@thijstriemstra removed owner|
|-|-|

Replying to [biny](#note_3):
> There isn't any change here. Only a new failing test.

Right, and that test can be added by supplying a diff.. I'll put the ticket up for review anyway.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

(In [27371]) Branching to 'image-fetch-specific-html-4049'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve commented|
|-|-|

(In [27373]) Branching to 'imap-fetch-specific-html-4049'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

I committed the test in the branch.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [27553]) Recognize some non-range values and avoid treating them as a range value

refs [#4049](https://github.com/twisted/twisted/issues/4049)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

r27553 is a terrible hack which partially resolves this issue.  I'm also attaching a patch which randomly changes a bunch of things around: this is the result of quite a few hours of trying to fix the parser for real.  My conclusion at this point is that there are too many hacks stacked on top of each other in this code.  See [#4124](https://github.com/twisted/twisted/issues/4124).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph set owner to @exarkun|
|-|-|

This looks OK to me, except the test has no docstring.  Write one, then merge.  Also, make it a good one.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set status to `closed`|
|-|-|

(In [27670]) Merge imap-fetch-specific-html-4049 

Author: biny, exarkun
Reviewer: glyph
Fixes: [#4049](https://github.com/twisted/twisted/issues/4049)

Fix a bug in the IMAP4 client parsing code which would misinterpret many html-formatted
message bodies as instead being part of the surrounding protocol bits and thus fail to
actually make the body content available in the result.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
